pkg/apis/seedmanagement/v1alpha1/types_managedseed.go

```diff
@@ -0,0 +1,216 @@
+	// Specification of the desired behavior of the Seed.
+	// +optional
+	Spec gardencorev1beta1.SeedSpec `json:"spec,omitempty" protobuf:"bytes,2,opt,name=spec"`
+}
+
+// Gardenlet specifies gardenlet deployment parameters and the GardenletConfiguration used to configure gardenlet.
+type Gardenlet struct {
+	// Deployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+	// the image, which bootstrap mechanism to use (bootstrap token / service account), etc.
+	// +optional
+	Deployment *GardenletDeployment `json:"deployment,omitempty" protobuf:"bytes,1,opt,name=deployment"`
+	// Config is the GardenletConfiguration used to configure gardenlet.
+	// +optional
+	Config *gardenletconfigv1alpha1.GardenletConfiguration `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`
+	// Bootstrap is the mechanism that should be used for bootstrapping gardenlet connection to the Garden cluster. One of ServiceAccount, Token.
+	// If specified, a service account or a bootstrap token will be created in the garden cluster and used to compute the bootstrap kubeconfig.
+	// If not specified, the gardenClientConnection.kubeconfig field will be used to connect to the Garden cluster.
+	// +optional
+	GardenConnectionBootstrap *GardenConnectionBootstrap `json:"gardenConnectionBootstrap,omitempty" protobuf:"bytes,3,opt,name=gardenConnectionBootstrap"`
+	// SeedConnection is the mechanism for gardenlet connection to the Seed cluster. Must equal ServiceAccount if specified.
+	// If not specified, the seedClientConnection.kubeconfig field will be used to connect to the Seed cluster.
+	// +optional
+	SeedConnection *SeedConnection `json:"seedConnection,omitempty" protobuf:"bytes,4,opt,name=seedConnection"`
+	// MergeParentConfig specifies whether the deployment parameters and GardenletConfiguration of the parent gardenlet
+	// should be merged with the specified deployment parameters and GardenletConfiguration. Defaults to false.
+	// +optional
+	MergeParentConfig bool `json:"mergeParentConfig,omitempty" protobuf:"varint,5,opt,name=mergeParentConfig"`
+}
+
+// GardenletDeployment specifies certain gardenlet deployment parameters, such as the number of replicas,
+// the image, which bootstrap mechanism to use (bootstrap token / service account), etc.
+type GardenletDeployment struct {
+	// ReplicaCount is the number of gardenlet replicas. Defaults to 1.
+	// +optional
+	ReplicaCount *int32 `json:"replicaCount,omitempty" protobuf:"varint,1,opt,name=replicaCount"`
+	// RevisionHistoryLimit is the number of old gardenlet ReplicaSets to retain to allow rollback. Defaults to 10.
+	// +optional
+	RevisionHistoryLimit *int32 `json:"revisionHistoryLimit,omitempty" protobuf:"varint,2,opt,name=revisionHistoryLimit"`
+	// ServiceAccountName is the name of the ServiceAccount to use to run gardenlet pods.
+	// +optional
+	ServiceAccountName string `json:"serviceAccountName,omitempty" protobuf:"bytes,3,opt,name=serviceAccountName"`
```

Comment:
Optional strings should be of type `*string`, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required
```suggestion
	ServiceAccountName *string `json:"serviceAccountName,omitempty" protobuf:"bytes,3,opt,name=serviceAccountName"`
```

Reply:
Hm, see my replies to @rfranzke comments earlier, and see also e.g. https://github.com/kubernetes/api/blob/master/core/v1/types.go for many examples of the contrary. Pointers are messier to deal with and I would in principle use them only if really needed. On the other hand the document you refer to is very explicit that this **should** be done this way, so I guess it should be done if we want to be compliant.

Reply:
As written above, I think some things in `core/v1` are legacy introduced before the official conventions were agreed on.

Reply:
Yes, will adapt all such code to make it compliant, see above.